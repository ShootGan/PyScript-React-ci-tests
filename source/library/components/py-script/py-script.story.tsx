import type { Meta, StoryFn } from "@storybook/react";
import PyScriptProvider from "~components/py-script-provider/py-script-provider";
import PyScript from "./py-script";
import type { PyScriptProperties, PyScriptTag } from "./py-script.types";

export default {
  component: PyScript,
} as Meta<typeof PyScript>;

const Template: StoryFn<PyScriptTag> = <T extends object>({
  ...rest
}: PyScriptProperties<T>): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript {...rest} />
    </PyScriptProvider>
  );
};

const LegacyTemplate: StoryFn<PyScriptTag> = <T extends object>({
  ...rest
}: PyScriptProperties<T>): JSX.Element => {
  return (
    <PyScriptProvider
      jsSource="https://pyscript.net/alpha/pyscript.js"
      cssSource="https://pyscript.net/alpha/pyscript.css"
    >
      <PyScript {...rest} />
    </PyScriptProvider>
  );
};

export const PyScriptLegacyHelloWorldExample: StoryFn<typeof PyScript> =
  LegacyTemplate.bind({});

PyScriptLegacyHelloWorldExample.args = {
  children: 'print("Hello world!")',
};

export const PyScriptHelloWorldExample: StoryFn<typeof PyScript> =
  Template.bind({});

PyScriptHelloWorldExample.args = {
  children: 'display("Hello world!")',
};

export const PyScriptFoliumExample: StoryFn<typeof PyScript> = Template.bind(
  {},
);

PyScriptFoliumExample.args = {
  source: "/folium_map.py",
  output: "folium",
  generateOutputTag: true,
  pyConfigProps: {
    type: "json",
    packages: new Set(["folium"]),
  },
};

export const PyScriptFoliumWithTomlConfigExample: StoryFn<typeof PyScript> =
  Template.bind({});

PyScriptFoliumExample.args = {
  source: "/folium_map.py",
  output: "folium",
  generateOutputTag: true,
  pyConfigProps: {
    type: "toml",
  },
  pyConfigContent: 'packages = ["folium"]',
};

export const PyScriptFoliumWithExternalTomlConfigExample: StoryFn<
  typeof PyScript
> = Template.bind({});

PyScriptFoliumWithExternalTomlConfigExample.args = {
  source: "/folium_map.py",
  output: "folium",
  generateOutputTag: true,
  pyConfigProps: {
    type: "toml",
    source: "./folium_map_config.toml",
  },
};

export const PyScriptMatplotlibExample: StoryFn<typeof PyScript> =
  Template.bind({});

PyScriptMatplotlibExample.args = {
  source: "/matplot_chart.py",
  output: "mpl",
  generateOutputTag: true,
  pyConfigProps: {
    type: "json",
    packages: new Set(["matplotlib", "numpy"]),
  },
};

export const PyScriptFoliumWithExternalJsonConfigExample: StoryFn<
  typeof PyScript
> = Template.bind({});

PyScriptFoliumWithExternalJsonConfigExample.args = {
  source: "/folium_map.py",
  output: "folium",
  generateOutputTag: true,
  pyConfigProps: {
    type: "json",
    source: "./folium_map_config.json",
  },
};

export const PyScriptLegacyFoliumExample: StoryFn<typeof PyScript> =
  LegacyTemplate.bind({});

PyScriptLegacyFoliumExample.args = {
  source: "/legacy_folium_map.py",
  output: "folium",
  generateOutputTag: true,
  pyEnvContent: new Set(["folium"]),
};

export const PyScriptLegacyFibonacciExample: StoryFn<typeof PyScript> =
  LegacyTemplate.bind({});

PyScriptLegacyFibonacciExample.args = {
  source: "/fibonacci.py",
};
