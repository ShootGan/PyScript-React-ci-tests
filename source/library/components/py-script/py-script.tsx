import propTypes from "prop-types";
import {
  DetailedHTMLProps,
  HTMLAttributes,
  WeakValidationMap,
  createElement,
  useMemo,
} from "react";
import {
  PyConfigPropertiesWithJsonType,
  PyConfigPropertiesWithTomlType,
  PyConfigPropertiesWithoutChildren,
} from "~components/py-config/py-config.types";
import ReactElementProps from "~root/source/library/types/react-element-properties/react-element-properties";
import PyConfig from "../py-config/py-config";
import PyEnv from "../py-env/py-env";
import type {
  PyScriptProperties,
  PyScriptPropertiesWithoutGeneric,
  PyScriptTag,
} from "./py-script.types";

/**
 * Allow to embed python scripts.
 * @type {PyScriptTag}
 * @returns {JSX.Element}
 * @param {string} properties.children - Python script content.
 * @param {string} properties.output - Id of element where output should be displayed. Works only with legacy version of PyScript.
 * @param {string} properties.source - Path to python file.
 * @param {string} properties.src - Alias for source.
 * @param {string} properties.generateOutputTag - If the value is true, it generates a `div` element. When the provided value is a string, it generates an element based on this value. If we want to limit only to real tags in HTML, we can check if the value meets the assumptions of the `keyof ReactHTML` type.
 * @param {string | string[] | Set<string> | ChildrenAsObject} properties.pyEnvContent - py-env children (deprecated - works only with legacy pyscript).
 * @param {Omit<PyEnvPropertiesBase, "children">} properties.pyEnvProps - py-env properties (deprecated - works only with legacy pyscript).
 * @param {string | (Record<string, any> & PyConfigSharedJsonProperties)} properties.pyConfigContent - py-config children
 * @param {Omit<ReactElementProps<DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>>,"children"> & (Omit<PyConfigPropertiesWithJsonType, "children"> | Omit<PyConfigPropertiesWithTomlType, "children"> | PyConfigPropertiesWithoutChildren)} properties.pyConfigProps - py-config properties
 * @see {@link https://docs.pyscript.net/latest/reference/elements/py-script.html} Original py-script element documentation.
 * @see {@link https://pyscript-react.github.io/} Pyscript-react element documentation.
 */
const PyScript: PyScriptTag = <T extends object>({
  children,
  output,
  generateOutputTag,
  pyEnvContent,
  pyEnvProps,
  pyConfigContent,
  pyConfigProps,
  src,
  source,
  ...rest
}: PyScriptProperties<T>): JSX.Element => {
  const pyConfigElement: "" | JSX.Element | undefined = useMemo(():
    | ""
    | JSX.Element
    | undefined => {
    const isPyConfigSource: boolean = Boolean(
      (pyConfigProps as Partial<PyConfigPropertiesWithoutChildren>)?.source,
    );
    const isPyConfigContentJsonType: boolean = pyConfigProps?.type === "json";
    const pyConfigElementWithContent: "" | JSX.Element | undefined =
      isPyConfigContentJsonType ? (
        <PyConfig
          {...(pyConfigProps as Omit<
            PyConfigPropertiesWithJsonType,
            "children"
          > &
            Omit<
              ReactElementProps<
                DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
              >,
              "children"
            >)}
        >
          {pyConfigContent as PyConfigPropertiesWithJsonType["children"]}
        </PyConfig>
      ) : (
        pyConfigContent && (
          <PyConfig
            {...(pyConfigProps as Omit<
              PyConfigPropertiesWithTomlType,
              "children"
            > &
              Omit<
                ReactElementProps<
                  DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
                >,
                "children"
              >)}
          >
            {pyConfigContent as PyConfigPropertiesWithTomlType["children"]}
          </PyConfig>
        )
      );
    return isPyConfigSource ? (
      <PyConfig {...(pyConfigProps as PyConfigPropertiesWithoutChildren)} />
    ) : (
      pyConfigElementWithContent
    );
  }, [pyConfigProps, pyConfigContent]);
  return (
    <>
      {pyConfigElement}
      {pyEnvContent && <PyEnv {...pyEnvProps}>{pyEnvContent}</PyEnv>}
      {output &&
        generateOutputTag &&
        createElement(
          typeof generateOutputTag === "string" ? generateOutputTag : "div",
          {
            id: output,
          },
        )}
      <py-script {...rest} src={source || src} output={output}>
        {children || ""}
      </py-script>
    </>
  );
};

PyScript.propTypes = {
  children: propTypes.string,
  output: propTypes.string,
  generateOutputTag: propTypes.oneOfType([propTypes.string, propTypes.bool]),
  pyEnvContent: propTypes.oneOfType([
    propTypes.string,
    propTypes.array,
    propTypes.exact({
      items: propTypes.arrayOf(propTypes.string),
      paths: propTypes.arrayOf(propTypes.string),
    }),
  ]),
  pyEnvProps: propTypes.object,
  pyConfigContent: propTypes.oneOfType([
    propTypes.string.isRequired,
    propTypes.object,
  ]),
  pyConfigProps: propTypes.object,
  src: propTypes.string,
  source: propTypes.string,
} as WeakValidationMap<PyScriptPropertiesWithoutGeneric>;

export default PyScript;
