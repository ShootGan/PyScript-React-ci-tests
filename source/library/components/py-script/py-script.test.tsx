import { cleanup, render, RenderResult } from "@testing-library/react";
import { FC } from "react";
import { expectType } from "tsd";
import PyScriptProvider from "~components/py-script-provider/py-script-provider";
import PyScript from "~components/py-script/py-script";
import { PyScriptTag } from "~components/py-script/py-script.types";

afterEach(cleanup);

const PyScriptHelloWorldComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript
        aria-label="Hello World generated based on Python code"
        data-testid="py-script"
      >
        print("Hello world!")
      </PyScript>
    </PyScriptProvider>
  );
};

const PyScriptFoliumMapComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript
        aria-label="Folium map generated based on Python code"
        data-testid="py-script"
        output="folium"
        generateOutputTag
        pyConfigProps={{
          type: "json",
        }}
        pyConfigContent={{
          packages: ["folium"],
        }}
        source="./test_data/folium_map.py"
      />
    </PyScriptProvider>
  );
};

const PyScriptFoliumMapWithExternalTomlConfigComponent: FC =
  (): JSX.Element => {
    return (
      <PyScriptProvider>
        <PyScript
          aria-label="Folium map generated based on Python code"
          data-testid="py-script"
          output="folium"
          generateOutputTag
          pyConfigProps={{
            source: "./test_data/folium_map_config.toml",
            type: "toml",
          }}
          source="./test_data/folium_map.py"
        />
      </PyScriptProvider>
    );
  };

const PyScriptLegacyFoliumMapComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript
        aria-label="Folium map generated based on Python code"
        data-testid="py-script"
        output="folium"
        generateOutputTag
        pyEnvContent={["folium"]}
        source="./test_data/folium_map.py"
      />
    </PyScriptProvider>
  );
};

const PyScriptLegacyFoliumMapWithGenerateOutputTagWithStringValueComponent: FC =
  (): JSX.Element => {
    return (
      <PyScriptProvider>
        <PyScript
          aria-label="Folium map generated based on Python code"
          data-testid="py-script"
          output="folium"
          generateOutputTag="section"
          pyEnvContent={["folium"]}
          source="./test_data/folium_map.py"
        />
      </PyScriptProvider>
    );
  };

describe("PyScript", (): void => {
  it("Renders component properly", async (): Promise<void> => {
    const { container }: RenderResult = render(<PyScriptHelloWorldComponent />);
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("Renders component properly with toml external config", async (): Promise<void> => {
    const { container }: RenderResult = render(
      <PyScriptFoliumMapWithExternalTomlConfigComponent />,
    );
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("Compile component with Python properly", (): void => {
    const { getByTestId }: RenderResult = render(
      <PyScriptHelloWorldComponent />,
    );
    // eslint-disable-next-line testing-library/prefer-screen-queries
    const pyScript: HTMLElement = getByTestId("py-script");
    setTimeout((): void => {
      expect(pyScript).toBeInTheDocument();
      expect(pyScript).toContainEqual("Hello world!");
    }, 5_000);
  });
  it("Renders component properly with output, source, pyEnvContent and generateOutputTag", async (): Promise<void> => {
    const { container }: RenderResult = render(
      <PyScriptLegacyFoliumMapComponent />,
    );
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("Compile component with Python folium example properly", (): void => {
    const { getByTestId }: RenderResult = render(
      <PyScriptLegacyFoliumMapComponent />,
    );
    // eslint-disable-next-line testing-library/prefer-screen-queries
    const pyScript: HTMLElement = getByTestId("py-script");
    setTimeout((): void => {
      expect(pyScript.getElementsByTagName("img").length).not.toBe(0);
    }, 5_000);
  });
  it('Generates a valid html tag on value: "true" for generateOutputTag', (): void => {
    const { container }: RenderResult = render(
      <PyScriptLegacyFoliumMapComponent />,
    );
    setTimeout((): void => {
      expect(container.getElementsByTagName("div")[0].id).toBe("folium");
    }, 5_000);
  });
  it('Generates a valid html tag on value: "keyof ReactHTML" for generateOutputTag', (): void => {
    const { container }: RenderResult = render(
      <PyScriptLegacyFoliumMapWithGenerateOutputTagWithStringValueComponent />,
    );
    setTimeout((): void => {
      expect(container.getElementsByTagName("section")[0].id).toBe("folium");
    }, 5_000);
  });
  it("Renders component properly with output, source, pyConfigContent and generateOutputTag", async (): Promise<void> => {
    const { container }: RenderResult = render(<PyScriptFoliumMapComponent />);
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("Has correct type", (): void => {
    expectType<PyScriptTag>(PyScript);
  });
});
