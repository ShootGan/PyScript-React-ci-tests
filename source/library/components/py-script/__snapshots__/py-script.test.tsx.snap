// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PyScript Renders component properly 1`] = `
<div>
  <py-script
    aria-label="Hello World generated based on Python code"
    data-testid="py-script"
  >
    print("Hello world!")
  </py-script>
</div>
`;

exports[`PyScript Renders component properly with output, source, pyConfigContent and generateOutputTag 1`] = `
<div>
  <py-config
    type="json"
  >
    {"packages":["folium"]}
  </py-config>
  <div
    id="folium"
  />
  <py-script
    aria-label="Folium map generated based on Python code"
    data-testid="py-script"
    output="folium"
    src="./test_data/folium_map.py"
  />
</div>
`;

exports[`PyScript Renders component properly with output, source, pyEnvContent and generateOutputTag 1`] = `
<div>
  <py-env>
    - folium
  </py-env>
  <div
    id="folium"
  />
  <py-script
    aria-label="Folium map generated based on Python code"
    data-testid="py-script"
    output="folium"
    src="./test_data/folium_map.py"
  />
</div>
`;

exports[`PyScript Renders component properly with toml external config 1`] = `
<div>
  <py-config
    src="./test_data/folium_map_config.toml"
    type="toml"
  />
  <div
    id="folium"
  />
  <py-script
    aria-label="Folium map generated based on Python code"
    data-testid="py-script"
    output="folium"
    src="./test_data/folium_map.py"
  />
</div>
`;
